sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -728,17 +786,15 @@ public void output(OutputT output, Instant timestamp, BoundedWindow window) {
    * {@link #bindWatermark} should never be implemented.
    */
   private class BeamFnStateBinder implements StateBinder {
-    private final Map<StateKey.BagUserState, Object> stateObjectCache = new HashMap<>();
-
     @Override
     public <T> ValueState<T> bindValue(String id, StateSpec<ValueState<T>> spec, Coder<T> coder) {
-      return (ValueState<T>) stateObjectCache.computeIfAbsent(
+      return (ValueState<T>) stateKeyObjectCache.computeIfAbsent(
           createOrUseCachedBagUserStateKey(id),
-          new Function<StateKey.BagUserState, Object>() {
+          new Function<StateKey, Object>() {
         @Override
```

Comment:
<!--new_thread; commit:c3d73c64facc4d4670f203a8fda5fafa1fb8ef8b; resolved:0-->
Nit: This indentation seems odd.